### PR TITLE
Refactor AST utilities

### LINKED ISSUE #1196: Standard way for extracting a node from the AST tree
Greetings,

SILE 0.10.15 on Ubuntu 20.04

In my package attempt (*), I tried to do something similar to the dockbook example in the SILE manual, where it uses `SILE.findInTree` to locate a command within the immediate children of the current tree, and then a `docbook.wipe` helper to nullify it before processing the content. 

In the manual's example, it works well, as it extracts a "title", processes it, and then _afterward_ wipes it and processes the content.

The problem in my case is that the extracted command is intended to be processed _after_ the content (basically, displayed below it), so I could not use the same trick. I ended up explicitly removing it from the tree, as follows (boilerplate removed)

```
omikhleia.extractFromTree = function (tree, command)
  for i=1, #tree do
    if type(tree[i]) == "table" and tree[i].command == command then
      return table.remove(tree, i)
    end
  end
end
```
It works, but:
* It doesn't feel right to do it in a package. It requires too much knowledge of the internal AST for what could be a fairly common use case (e.g. extracting a caption from a figure environment, etc.) - I have the same remark for the `docbook.wipe` helper, actually.
* This could (should, IMHO) be a standard command, in the same vein as `SILE.findInTree`
* If so, it could be made more robust and efficient, e.g.
  * It could check there is no extraneous command by that name remaining and issue a warning (the command may or may not exist in the broader scope of the document)
  * I am not sure `table.remove` is efficient if the processing tree gets big. If so implementation could differ (e.g. just flag the node as already processed / to be ignored later, but it then requires going deeper into the AST processing)  

EDIT (*) Just for the curious-minded, my [package attempt](https://github.com/Omikhleia/omikhleia-sile-packages/)

### PR DESCRIPTION
**Pull request by Omikhleia**
Proposal, motivated by #1866 
- Move existing AST utilities to `SU.ast`
  - _Rationale:_  While the SILE AST is just a table with some structure (tied to `SILE.process` etc.), having some convenience function to manipulate is good -- and a must-have for inputter and package designers. The existing utilities were however scattered (some at the root of `SU`, one as an inputter method, another in the inputfilter package, etc.)...
- Refactored as a separate utility file.
  - _Rationale:_  We have lots of different/unrelated stuff in `SU`, a better organization makes maintenance of similar functions easier and is the path for curating those utilities progressively. Later, we might at a glance easily check these if we ever want to change the AST (add new constructs, make it more object-oriented, etc.)
  - Adding in-code documentation: follows VSCode Lua extension syntax (I tried LDoc initially but could not make it work with VSCode... So there's nothing definitive here, I just thought something would be better than nothing.)
- Proposed additions (taken from my **silex.ast** layer)
   - `SU.ast.extractFromTree` - complements `SU.ast.findInTree`
      - Closes #1196 
   - `SU.ast.createCommand` - as in inputfilter, but in inputters etc. this is convenient (and we might not even have loaded any package yet, there, by nature)
   - `SU.ast.createStructuredCommand` - convenience when passing an array of nodes (avoiding useless nesting)
   - `SU.ast.trimSubContent` and `SU.ast.processAsStructure` - fairly useful esp. in XML parsing context, when the schema defines structured elements where text/spaces aren't expected (unless due do indentation, but to be ignored in the ouput).
- Other changes (also from **silex.ast**)
  - `SU.ast.subContent` is simplified and no longer includes a weird `id=stuff` entry - discussed in #1834 (items **B1** and **B2**)

### CODE REVIEW

**Comment by alerque** [comment]
> Adding in-code documentation: follows VSCode Lua extension syntax (I tried LDoc initially but could not make it work with VSCode... So there's nothing definitive here, I just thought something would be better than nothing.)

Something is definitely better than nothing. We'll likely be going the LDoc route eventually (c.f. #1337) but it will be much easier to adapt something than nothing!

File: core/utilities-ast.lua [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    end
+  end
+end
+
+--- Find and extract (remove) a command node in a SILE AST tree,
+--- looking only at the first level.
+---@param tree      table       AST tree
+---@param command   string      command name
+---@return          table|nil   AST command node
+function ast.extractFromTree (tree, command)
```
Comment (alerque) [inline core/utilities-ast.lua:24]: I don't think "extract" is the right verb here, at least to me that implies getting a node out in isolation, but also I would have assumed it *left* the original tree alone. This removes it from the tree, so "slice" or perhaps "extricate" might be a better name that suggests the original is being removed.

**Review by alerque — COMMENTED** [review]
This will need a `BREAKING CHANGE` log message, but I can take care of that. Overall I think it's looking good. Just doing some tinkering on my end to see what I think of the details.

File: core/utilities-ast.lua (hunk 1 above)
Comment (Omikhleia) [reply core/utilities-ast.lua]: maybe simpy `removeFromTree` (and the `@return` becomes "Removed AST command node") ?

**Comment by Omikhleia** [comment]
By the way @alerque I was about to ask whether we should actually split utilities in subfolders (rather than `-xxx` files), and I just see you felt alike :) :+1: 

Would that be ok if we go on splitting the utilities  in subfolders? Currently we still have in the main files
- Things pertaining to casting
- Things pertaining to error/warning etc.
- Things pertaining to general utilities (map, max, etc.)
- Things pertaining to unicode and characters
- A bunch of other things likely...

Perhaps starting to split them would allow us to properly group them by "nature" and document them progressively?

**Comment by Omikhleia** [comment]
> Perhaps starting to split them would allow us to properly group them by "nature" and document them progressively?

That could be another issue addressed later, obviously, just asking here as it sounds quite logical.

**Comment by alerque** [comment]
> Would that be ok if we go on splitting the utilities in subfolders?

Yes. Smaller files/modules with tighter scopes are almost always better for Lua in my opinion.

**Review by alerque — APPROVED** [review]

**Comment by alerque** [comment]
P.S. I'm also seriously scratching my head why we have "core" and "utilities". It seems to me the way our source organization is sorting out almost everything in "core" is just utilities. Basically anything that is just a function as opposed to a class than gets instantiated and maintains some sort of state as SILE runs.

**Comment by Omikhleia** [comment]
> P.S. I'm also seriously scratching my head why we have "core" and "utilities". It seems to me the way our source organization is sorting out almost everything in "core" is just utilities. Basically anything that is just a function as opposed to a class than gets instantiated and maintains some sort of state as SILE runs.

The line is blurry, indeed.
A few things stand apart, though and hint towards different natures:
- line breaker, page builder, hyphenator, fonts = core algorithm (some are classes or alike, actually): pretty specific to the field of digital typesetting
- color, units, nodes, frame... = low level common concepts and data structures
- cli, repl... = their name says it all, running on command line
- _most_ current "utilities" are things you one could work without
    - Many just avoid lots of code and/or make things uniform
    - Several could find their place in other pieces software actually (error handling with stack, string encoding or collation, etc.)

So there could be some different levels of concern and design here ;-)

**Pull request merged**